### PR TITLE
Better WFT buffer

### PR DESCRIPTION
**Pull request by Sushisource**
## What was changed
Better task buffering.

## Why?
To avoid timing out concurrent queries. There is no reason to buffer more than one WFT, speculative or not, for the reasons described in the second issue. We already definitely will never try to apply more than one WFT at a time if the split brain edge case mentioned there happens.

## Checklist

1. Closes https://github.com/temporalio/sdk-core/issues/578 and closes https://github.com/temporalio/sdk-core/issues/578

3. How was this tested:
New tests

4. Any docs updates needed?

### CODE REVIEW

File: core/src/worker/workflow/workflow_stream.rs [hunk 1]
```diff
@@ -31,6 +31,7 @@ use tracing::{Level, Span};
 pub(super) struct WFStream {
     runs: RunCache,
     /// Buffered polls for new runs which need a cache slot to open up before we can handle them
+    /// TODO: Does this even need to exist any more?
```
Comment (Sushisource) [inline core/src/worker/workflow/workflow_stream.rs:34]: Check this

File: core/src/worker/workflow/workflow_stream.rs [hunk 2]
```diff
@@ -339,48 +341,36 @@ impl WFStream {
...
-                res = self.instantiate_or_update(PermittedWFT {
-                    work: wft,
-                    permit: t.permit,
-                    paginator: pag,
-                });
+                .or_else(|| self.buffered_polls_need_cache_slot.pop_front());
+            if let Some(wft) = maybe_buffered_wft {
+                res = self.instantiate_or_update(wft);
+                // We accept that there might be tasks remaining in the buffer if we evicted and
+                // re-instantiated here. It's likely those tasks are now invalidated anyway.
```
Comment (Sushisource) [inline core/src/worker/workflow/workflow_stream.rs:373]: I could theoretically carry over any remaining buffered tasks to the next instantiation, but it adds quite a lot of complexity for little benefit IMO.

File: core/src/worker/workflow/workflow_stream.rs (hunk 1 above)
Comment (Sushisource) [reply core/src/worker/workflow/workflow_stream.rs]: It does, in fact, need to exist.

File: core/src/worker/workflow/managed_run.rs [hunk 3]
```diff
@@ -71,11 +73,13 @@ pub(super) struct ManagedRun {
...
-    /// run. This can happen when lang takes too long to complete a task and the task times out, for
-    /// example. Upon next completion, the buffered response will be removed and can be made ready
-    /// to be returned from polling
-    buffered_resp: Option<PermittedWFT>,
+    /// Contains buffered poll responses from the server that apply to this run. This can happen
+    /// when:
+    ///   * Lang takes too long to complete a task and the task times out
+    ///   * Many queries are submitted concurrently and reach this worker (in this case, multiple
+    ///     tasks can be outstanding)
+    ///   * Multiple speculative tasks (ex: for updates) may also exist at once
```
Comment (Quinn-With-Two-Ns) [inline core/src/worker/workflow/managed_run.rs:81]: Might be worth clarifying this statement that if multiple are sent only the latest is valid

File: core/src/worker/workflow/mod.rs [hunk 4]
```diff
@@ -1003,6 +1021,38 @@ enum WFTReportStatus {
...
+    fn has_tasks(&self) -> bool {
+        self.wft.is_some() || !self.query_only_tasks.is_empty()
+    }
+
+    /// Remove and return the next WFT from the buffer that should be applied
+    fn get_next_wft(&mut self) -> Option<PermittedWFT> {
+        self.wft
+            .take()
+            .or_else(|| self.query_only_tasks.pop_front())
+    }
```
Comment (dandavison) [inline core/src/worker/workflow/mod.rs:1054]: A comment about the ordering preference for WFT ahead of Query-only tasks would be good.

File: core/src/worker/workflow/managed_run.rs (hunk 3 above)
Comment (Sushisource) [reply core/src/worker/workflow/managed_run.rs:81]: Will add

File: core/src/worker/workflow/mod.rs (hunk 4 above)
Comment (Sushisource) [reply core/src/worker/workflow/mod.rs:1054]: Will add, and this reminds me I meant to test that we won't reply to queries that are "old" if we processed a newer WFT than they targeted.

**Comment by Sushisource** [comment]
It looks like I may *need* to carry over buffered tasks, or else tests that are validating small-cache scenarios have some issues.
